# Instructions for use
### Before starting the application
In order to use the application, one needs to download the source code of the application from [the most up to date release](https://github.com/MattiKannisto/ot-harjoitustyo/releases/tag/loppupalautus). Before starting to use the application, one needs to install its dependencies in the command line by changing to the root folder of the project and executing the command **poetry install**. The application uses a SQLite database which needs to be initialized prior to first use of the application. This can be done in command line by changing to the root folder of the project and executing command **poetry run invoke initialize**.

### Starting the application
The application can be started from the root folder of the project with command **poetry run invoke start**. This will open the login view:

![a login screen](https://github.com/MattiKannisto/ot-harjoitustyo/raw/master/dokumentaatio/kuvat/login_view)

One can create a user account by clicking the button 'Create user account' which will open a create account view:

![a register screen](https://github.com/MattiKannisto/ot-harjoitustyo/raw/master/dokumentaatio/kuvat/create_account_view)

In order to create an account to the application, one needs to give an unique username (at least 5 characters), password (between 10 and 30 characters) and a directory into which the output files of the application will be save. Please note that you cannot select a directory that another user is already using for the application or a directory to which you are not authorized to write files into. If you insert invalid information into any of the fields in this view, the application will notify you of this. The lack of writing rights to the output folder will be notified only when you actually try to generate the output files since your rights might have changed between creating the account and using the application. Once the account has been successfully created, the register screen will be closed and the login screen will be re-opened. If the user gives correct username and password (the application will notify the user if he/she gives incorrect username and/or password), the login screen will be closed and the main view of the application will be opened:

![the main screen](https://github.com/MattiKannisto/ot-harjoitustyo/raw/master/dokumentaatio/kuvat/main_view)

### Using the application
The application assumes that the user is familiar with basic molecular biology and knows the basics of designing sequencing primers. Briefly, the DNA fragments added to the application consist of a name and a nucleotide sequence. Nucleotide sequences of DNA consist of four nucleotides and the order of the nucleotides determine the function of the DNA fragment. For protein coding sequences, the nucleotide sequence is read in triplets of nucleoties, e.g. codons. Each codon codes for an amino acid, one codon codes for the start of translation of DNA sequence into protein sequence and two nucleotides code for stopping the translation. Different organisms might have different codons for different amino acids but this application uses genetic code of Homo sapiens. Sequencing primers are short fragments of DNA with an identical nucleotide sequence of the template DNA into which they will anneal to in the sequencing reaction. The length of this sequence and its GC content (percentage of nucleotides 'G' and 'C') affect successfulness of the sequencing reactions and it is thus important that these values can be adjusted. The application generates sequencing primers for both forward strand and reverse strand of the DNA fragment. For additional information, the user is adviced on using for example following search words in any internet search engines: "DNA", "protein", "genetic code", "central dogma of molecular biology" or "sequencing primer design".

The main view consist of entries and buttons for adding new DNA fragments and analyzing existing DNA fragments and of a notification area where information regarding the successfulness of different actions will be displayed to the user. Notifications of successful actions are displayed in green font, notifications that warn that something might not have been carried out with sufficient quality are displayed in yellow font and notifications that something could not be carried out at all are displayed in red font. The buttons for adding and processing DNA fragments will be activated only when the text box for adding a new DNA fragment are filled in or an existing DNA fragment is selected, respectively. When adding a new DNA fragment, the user is required to give a unique name for the fragment. The sequence of the fragment needs also contain only the letters representing naturally occurring nucleotides. If either of these conditions is not met, and also when a fragment can be successfully added, the user is notified of this in the black notification area. The DNA fragments (and user account, user account settings, and primers generated) added will be stored in the SQLite database and the user will be able to browse the DNA fragments he/she has already added to the system from the dropdown menu next to the text 'Now processing DNA fragment". The generated protein sequence and sequencing primer files will be stored in the folder specified by the user when creating the user account. This directory can be changed by selecting 'Options' and then 'Settings' which will open a new window showing settings. In this screen the user can also change settings for generating the sequencing primers and also to delete the user account. The user can log out from the application by clicking 'Options' and then 'Logout'. The application can be closed by clicking 'Options' and then 'Exit'.

### Additional instructions
If one needs help in using the application, one should click the menu 'Options' and then 'Help'. In login, settings and create user account views this will display instructions on the top row of the view. In main view that will be opened after loggin in to the application, doing this will activate help mode. In the help mode one can click any of the buttons or menu items and instructions on what they do outside the help mode will be displayed in the notification area of the view. By clicking 'Help' again in the 'Options' menu the help mode will be disabled and one can resume using the application. The application will further notify the user of invalid information entered or if some of the operations carried out for the DNA fragments are not successful.
